File: articles/calculus-for-systems-modeling/articles/gradient-divergence-and-curl/r/field_operator_audit.R
scalar_field <- function(x, y) {
  x^2 + y^2
}

vector_field <- function(x, y) {
  c(-y, x)
}

gradient_field <- function(x, y) {
  c(2 * x, 2 * y)
}

divergence_field <- function(x, y) {
  0
}

curl_2d <- function(x, y) {
  2
}

grid_values <- function(step) {
  seq(-1, 1, by = step)
}

audit_field_operators <- function(step, scenario) {
  values <- grid_values(step)
  grad_magnitudes <- c()
  divergences <- c()
  curls <- c()

  for (x in values) {
    for (y in values) {
      grad <- gradient_field(x, y)
      grad_magnitudes <- c(grad_magnitudes, sqrt(sum(grad^2)))
      divergences <- c(divergences, divergence_field(x, y))
      curls <- c(curls, curl_2d(x, y))
    }
  }

  warning <- ifelse(
    step > 0.5,
    "Grid step is coarse; local derivative structure may be undersampled.",
    "Synthetic field-operator audit; document field definitions, units, grid, and boundary rules."
  )

  data.frame(
    scenario = scenario,
    grid_step = step,
    point_count = length(values)^2,
    mean_gradient_magnitude = mean(grad_magnitudes),
    maximum_gradient_magnitude = max(grad_magnitudes),
    mean_divergence = mean(divergences),
    mean_curl = mean(curls),
    maximum_abs_curl = max(abs(curls)),
    field_description = "scalar f=x^2+y^2; vector F=<-y,x>",
    warning = warning
  )
}

results <- rbind(
  audit_field_operators(1.0, "coarse_grid"),
  audit_field_operators(0.5, "medium_grid"),
  audit_field_operators(0.25, "fine_grid")
)

dir.create("outputs/tables", recursive = TRUE, showWarnings = FALSE)
write.csv(results, "outputs/tables/r_field_operator_audit.csv", row.names = FALSE)
print(results)

File: articles/calculus-for-systems-modeling/articles/gradient-divergence-and-curl/julia/field_operator_audit.jl
scalar_field(x, y) = x*x + y*y
vector_field(x, y) = (-y, x)
gradient_field(x, y) = (2.0*x, 2.0*y)
divergence_field(x, y) = 0.0
curl_2d(x, y) = 2.0
vec_norm(v) = sqrt(v[1]^2 + v[2]^2)
grid_values(step) = collect(-1.0:step:1.0)

function audit_field_operators(step, scenario)
    values = grid_values(step)
    grad_magnitudes = Float64[]
    divergences = Float64[]
    curls = Float64[]

    for x in values
        for y in values
            push!(grad_magnitudes, vec_norm(gradient_field(x,y)))
            push!(divergences, divergence_field(x,y))
            push!(curls, curl_2d(x,y))
        end
    end

    warning = step > 0.5 ? "Grid step is coarse; local derivative structure may be undersampled." : "Synthetic field-operator audit; document field definitions, units, grid, and boundary rules."
    return scenario, step, length(values)^2, sum(grad_magnitudes)/length(grad_magnitudes), maximum(grad_magnitudes), sum(divergences)/length(divergences), sum(curls)/length(curls), maximum(abs.(curls)), "scalar f=x^2+y^2; vector F=<-y,x>", warning
end

println("scenario,grid_step,point_count,mean_gradient_magnitude,maximum_gradient_magnitude,mean_divergence,mean_curl,maximum_abs_curl,field_description,warning")
for case in [(1.0,"coarse_grid"),(0.5,"medium_grid"),(0.25,"fine_grid")]
    println(join(audit_field_operators(case[1], case[2]), ","))
end
